File: implementations/java/astar/Node.java
import java.util.ArrayList;

/**
 * Represents a standalone node of a graph.
 */
public class Node implements Comparable<Node> {
    private static int idCounter = 0;
    public int id;

    public Node parent = null;

    public ArrayList<Edge> neighbours;

    // Evaluation functions
    public double costOfNode = Double.MAX_VALUE;
    public double distanceToStartNode = Double.MAX_VALUE;

    public double heuristic; 

    public Node(double heuristic) {
        this.heuristic = heuristic;
        this.id = idCounter++;
        this.neighbours = new ArrayList<Edge>();
    }

    public void addEdge(Node node, int weight) {
        neighbours.add(new Edge(node, weight));
    }
    
    public double getHeuristic() {
        return this.heuristic;
    }

    public boolean equals(Node node) {
        return this.id == node.id;
    }
 
    @Override
    public int compareTo(Node node) {
        return Double.compare(this.costOfNode, node.costOfNode);
    }
}

File: implementations/go/astar/Node.go
package main

import (
	"fmt"
	"math"
)

// Represents a standalone node of a graph.
//
// params:
//  - id {int} The unique identifier of the node.
//  - heuristic {int} The heuristic of the node.
type Node struct {
	id                  int
	index				int
	heuristic           int
	parent              *Node
	neighbours          []*Edge
	costOfNode          int
	distanceToStartNode int
}

func CreateNode(id int, heuristic int) *Node {
	node := new(Node)
	node.id = id
	node.index = 0
	node.heuristic = heuristic
	node.parent = nil
	node.neighbours = []*Edge{}
	node.costOfNode = math.MaxInt
	node.distanceToStartNode = math.MaxInt

	return node
}

func (node *Node) GetHeuristic() int {
	return node.heuristic
}

func (node *Node) AddEdge(target Node, weight int) {
	edge := new(Edge)
	edge.weight = weight
	edge.source = node
	edge.target = &target
	node.neighbours = append(node.neighbours, edge)
}

func (node *Node) Equals(n *Node) bool {
	return node.id == n.id
}

func (node Node) String() string {
	return fmt.Sprintf("%d (%d) costs %d and is %d units from the source", node.id, node.heuristic, node.costOfNode, node.distanceToStartNode)
}

func (node Node) PrintNeighbours() {
	fmt.Print("[")
	for _, edge := range node.neighbours {
		fmt.Print(edge.target.id, " ")
	}
	fmt.Println("]")
}
